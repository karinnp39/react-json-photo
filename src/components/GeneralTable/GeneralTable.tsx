import React from 'react';
import { get } from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import TableConfigType from '../../types/TableConfigType';
import PhotoType from '../../types/Photo';

interface PhotosProps {
  pending: boolean
  items: Array<PhotoType>
  error: boolean
  tableConfig: TableConfigType
  captionText: string
}


const Photos: React.FC<PhotosProps> = (props) => {
  
  const { items, pending, error, tableConfig, captionText } = props;

  let displayItems: (PhotoType)[] = items;

  const noItems = items? (items.length <= 0) : true;
  const tableHeaders = tableConfig.headers;
  const alignments = tableConfig.alignments
  const cols = tableConfig.cols;
  const getFieldValue = (item: PhotoType, header:   any) => {
    return get(item, header.field);
  }
  const getFieldPhoto = (item: PhotoType, header:   any) => {
    return item.thumbnailUrl;
  }

  return(
    <>
    {!error && !pending && !noItems &&
    (<div className="table-responsive">
      <table className="table w-100 d-block d-md-table table-striped table-hover center">
      <thead className="thead">
        <tr className="row d-flex align-items-stretch">
        {!pending && !error && 
         tableHeaders.map( (header, i) => {
          let col = (cols[i]===-1) ? "col" : `col-${cols[i]}`;
          return(
          <th
            className={`${col} text-center`}
            id={`${header.field}`}
            key={`${header.field}`}>{header.display}
          </th>)
        }
        )}
        </tr>
      </thead>
      <tbody>
      {!pending && !error && items &&      
        displayItems
        .map(item => (
          <tr
            className="row"
            id={`item.id`}
            key={item.title}
          >
            {tableHeaders
            .map((header, i:number) => {
              let col = (cols[i]===-1) ? "col" : `col-${cols[i]}`;
              if (header.display === "Photo") {
                return(
                  <td 
                  key={`${item.id}_h_${i}`} 
                  className={`text-${alignments[i]} ${col}`}  
                  >
                    <img src={getFieldPhoto(item, header)} alt="avatar"/>
                  </td>)
              } else {
                return(
                <td 
                key={`${item.id}_h_${i}`} 
                className={`text-${alignments[i]} ${col}`}  
                >{getFieldValue(item, header)}
                  </td>)
              }
              
            })}
          </tr>
        ))
      }
      </tbody>
    </table></div>
    )}
    <hr />
    {!error && !pending && (
        !noItems? (
        <h3 className="caption">{displayItems.length} {captionText}  </h3>) :
          (<h3 className="caption">No {captionText} found</h3>)
    )}
    {error && 
    <div className="container">
      <h2>Error message</h2>
    </div>}
    {pending && 
    <div className="container">
      <FontAwesomeIcon icon="spinner" size="3x" spin /></div>}
    </>
  );
};

export default Photos;