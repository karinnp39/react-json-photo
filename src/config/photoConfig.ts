export const photoConfig = {
  "apiUrl": "https://jsonplaceholder.typicode.com/photos",
  "headers": [ 
    { "field": "thumbnailUrl", "display": "Photo" },
    { "field": "title", "display": "Title" }, 
  ],
  "alignments": [ "center", "left" ],
  "cols": [ 4, 8 ],
  "title": "Photos",
  "searchCaseSensitive": false
};

export default photoConfig;
