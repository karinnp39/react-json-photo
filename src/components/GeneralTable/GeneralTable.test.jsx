import React from 'react';
import { shallow, unmount } from 'enzyme';
import '@testing-library/jest-dom/extend-expect'
import {render, fireEvent} from '@testing-library/react';

import GeneralTable from './GeneralTable';
import { photoConfig as PhotoTableConfig } from '../../config/photoConfig';
import testPhoto from '../../utils/testPhoto';

const historyMock = jest.fn(() => {
  return true;
});

let photoTableProps = {
  items: testPhoto,
  tableConfig: {...PhotoTableConfig},
  pending: false,
  error: false,
  onRowClick: historyMock,
  captionText: "Photos"
};

describe('enzyme smoke tests', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<GeneralTable 
      {...photoTableProps} />);
  });

  afterAll(() => {
    unmount(wrapper);
  });
  it('matches napshot', () => {   
    expect(wrapper).toMatchSnapshot();
  });
  
  it('actually have a single table element', () => {
    expect(wrapper.find('table').length).toBe(1);
  });
});

describe('the initial test photo data', () => {
it('should display a the test photo in table row', () => {
    const { container, getByText } = render(<GeneralTable {...photoTableProps} />);

    const tableNode = container.querySelector('table');
    expect(tableNode).toBeInTheDocument();

    const tdNode = getByText(testPhoto[0].name);
    expect(tdNode.textContent).toBe('Ervin Howell');
  });

  it('should display the photo table configuration columns', () => {
    const { container, getByText } = render(<GeneralTable {...photoTableProps} />);

    const headerNode = container.querySelector('thead');
    expect(headerNode).toBeInTheDocument();

    PhotoTableConfig.headers.forEach((value, key) => {
      const tdNode = getByText(value['display']);
      expect(tdNode.textContent).toBe(PhotoTableConfig.headers[key].display);
    });
  });

  it('should display the correct number of photos', () => {
    const { container } = render(<GeneralTable {...photoTableProps} />);
    const captionNode = container.querySelector('.caption');
    expect(captionNode).toBeInTheDocument();

    const lengthStr = testPhoto.length.toString();
    expect(captionNode.textContent.includes(lengthStr)).toBeTruthy();
  });
});

describe('The photo table\'s row click', () => {
  it('calls correct function from parent component', async () => {
    const { container } = render(<GeneralTable {...photoTableProps} />);

    const tableRow = container.querySelector('tbody tr');
    fireEvent.click(tableRow);
    await expect(historyMock).toHaveBeenCalled();
  });
});
