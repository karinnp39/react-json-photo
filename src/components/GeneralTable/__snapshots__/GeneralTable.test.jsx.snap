// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme smoke tests matches napshot 1`] = `ShallowWrapper {}`;
